walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,154 @@
+                                                ArrayList<TimeRange> allTimesRequiredAttendees,
+                                                ArrayList<TimeRange> notAvailableTimes,
+                                                ArrayList<TimeRange> availableTimes,
+                                              ) {
+    TimeRange current = allTimesRequiredAttendees.get(0);
+    for (TimeRange time : allTimesRequiredAttendees) {
+      // Checks whether our attendees have overlapping TimeRanges.
+      if (time.overlaps(current)) {
+        int start = current.start();
+        current = TimeRange.fromStartEnd(start, time.end() < current.end() ? current.end() : time.end(), false);
+      } else {
+        // Adds TimeRange when they don't conflict with each other.
+        notAvailableTimes.add(current);
+        current = time;
+        }
+    }
+    // Adding last one, we now have the full calendar of not available time ranges. 
+    notAvailableTimes.add(current);
+    int numberOfTimes = notAvailableTimes.size();
+    TimeRange first = notAvailableTimes.get(0);
+    TimeRange last = notAvailableTimes.get(numberOfTimes - 1);
+    
+    // Creates TimeRanges that include the start/end of the day.
+    TimeRange firstHalf = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, first.start(), false);
+    TimeRange secondHalf = TimeRange.fromStartEnd(last.end(), TimeRange.END_OF_DAY, true);
+    
+    // Takes out TimeRanges that do not last enough time. 
+    if (firstHalf.duration() >= request.duration()) {
+      availableTimes.add(firstHalf);
+    }
+    if (secondHalf.duration() >= request.duration()) {
+      availableTimes.add(secondHalf);
+    }
+    // Checks whether we have enough TimeRanges to iterate through
+    if (numberOfTimes > 1) {
+      for (int j = 0; j < numberOfTimes - 1; j++) {
+        TimeRange newTime = TimeRange.fromStartEnd(notAvailableTimes.get(j).end(), notAvailableTimes.get(j + 1).start(), false);
+        if (newTime.duration() >= request.duration()) {
+          availableTimes.add(newTime);
+        }
+      }
+    }
+    return availableTimes;                              
+  }
+
+  /**
+  * Takes in a collection of Events and a MeetingRequest and is expected to return
+  * a collection of TimeRanges in which the MeetingRequest can be satisfied given the
+  * the constraints posed by the attendees' events. 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<String> attendeesWithoutEvents = new ArrayList<>(request.getAttendees());
+    int initialAttendeeCount = attendeesWithoutEvents.size();
+    ArrayList<String> optionalAttendeesWithoutEvents = new ArrayList<>(request.getOptionalAttendees());
+    int initialOptionalAttendeeCount = optionalAttendeesWithoutEvents.size();
+    
+    // Stores all TimeRanges of mandatory attendees with no repetitions.
```

Comment:
Thanks for inserting these explanatory comments! They've been pretty useful to me in a few places. I should note that ideally the variably names are descriptive enough that comments aren't required. I think that's possible here but I'm not going to stress it in this review since variable naming is definitely a learned skill. Something to keep in mind for the capstone project : )